contracts/BC_fusion/StakeCredit.sol

```diff
@@ -0,0 +1,286 @@
+    error ZeroAmount();
+    error InsufficientBalance();
+    error NoUnbondRequest();
+    error NoClaimableUnbondRequest();
+
+    /*----------------- storage -----------------*/
+    address public validator; // validator's operator address
+    uint256 public totalPooledBNB; // total reward plus total BNB staked in the pool
+
+    // hash of the unbond request => unbond request
+    mapping(bytes32 => UnbondRequest) private _unbondRequests;
+    // delegator address => unbond request queue(hash of the request)
+    mapping(address => DoubleEndedQueueUpgradeable.Bytes32Deque) private _unbondRequestsQueue;
+    // delegator address => personal unbond sequence
+    mapping(address => CountersUpgradeable.Counter) private _unbondSequence;
+
+    /*----------------- structs and events -----------------*/
+    struct UnbondRequest {
+        uint256 shares;
+        uint256 bnbAmount;
+        uint256 unlockTime;
+    }
+
+    event RewardReceived(uint256 rewardToAll, uint256 commission);
+
+    /**
+     * @notice only accept BNB from `StakeHub`
+     */
+    receive() external payable onlyStakeHub {
+        totalPooledBNB += msg.value;
+    }
+
+    /*----------------- init -----------------*/
+    /*
+     * @param _validator validator's operator address
+     * @param _moniker validator's moniker
+     */
+    function initialize(address _validator, string calldata _moniker) external payable initializer onlyStakeHub {
+        string memory name_ = string.concat("Stake ", _moniker, " Credit");
+        string memory symbol_ = string.concat("st", _moniker);
+        __ERC20_init_unchained(name_, symbol_);
+
+        validator = _validator;
+
+        _bootstrapInitialHolder(msg.value);
+    }
+
+    /*----------------- external functions -----------------*/
+    /**
+     * @param delegator the address of the delegator
+     * @return shares the amount of shares minted
+     */
+    function delegate(address delegator) external payable onlyStakeHub returns (uint256 shares) {
+        if (msg.value == 0) revert ZeroAmount();
+        shares = _mintAndSync(delegator, msg.value);
+    }
+
+    /**
+     * @param delegator the address of the delegator
+     * @param shares the amount of shares to be undelegated
+     * @return bnbAmount the amount of BNB to be unlocked
+     */
+    function undelegate(address delegator, uint256 shares) external onlyStakeHub returns (uint256 bnbAmount) {
+        if (shares == 0) revert ZeroAmount();
+        if (shares > balanceOf(delegator)) revert InsufficientBalance();
+
+        // add to the queue
+        bnbAmount = _burnAndSync(delegator, shares);
+        uint256 unlockTime = block.timestamp + IStakeHub(STAKE_HUB_ADDR).unbondPeriod();
+        UnbondRequest memory request = UnbondRequest({ shares: shares, bnbAmount: bnbAmount, unlockTime: unlockTime });
+        bytes32 hash = keccak256(abi.encodePacked(delegator, _useSequence(delegator)));
+        _unbondRequests[hash] = request;
```

Comment:
I would like to enhance the security further: 
Check the `_unbondRequests[hash]` is not empty before assigning it, and do not delete _unbondRequests[hash] after claiming it.   (Just in case there are any replay attack)

Reply:
Done